### PR TITLE
[SPARK-11027][SQL] Better group distinct columns in query compilation for aggregation query

### PR DESCRIPTION
**Pull request by viirya**
JIRA: https://issues.apache.org/jira/browse/SPARK-11027

Currently, for a query like:

```
SELECT sum(distinct value1), kEY - 100, count(distinct value1) FROM agg2 GROUP BY Key - 100
```

We will treat it as having two distinct columns because sum is working on a cast of value1. We can ignore cast expression for such cases to have just one distinct column.

### CODE REVIEW

**Comment by JoshRosen** [comment]
One corner-case that I'm curious about: is it possible to have cases where the relative ordering of the distinct and cast operations can affect the answer? I'm thinking about a case where you have two values which are unequal but which become equal after casting, such as casting two floats to integers (which involves a loss of precision).

My understanding of our current casting semantics (please correct me if I'm wrong!) is that our automatic/implicit casts will only serve to widen the type without a loss of precision. As a result, it might be fine to push the cast after the distinct if it's an implicit cast. But what about explicit casts? If a user chose to cast a double to an int inside of their `sum`, e.g. `sum(distinct int(foo))`, then would this optimization be unsafe?

/cc @yhuai, who originally filed this JIRA.

**Comment by JoshRosen** [comment]
By the way, if we do decide to move forward with this optimization then I think that this PR needs planner tests to make sure that the optimization is being applied properly.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/utils.scala [hunk 1]
```diff
@@ -166,7 +166,7 @@ object Utils {
     // It is safe to call head at here since functionsWithDistinct has at least one
     // AggregateExpression2.
     val distinctColumnExpressions =
-      functionsWithDistinct.head.aggregateFunction.children
+      functionsWithDistinct.flatMap(_.aggregateFunction.children)
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/utils.scala]: Why is this change necessary? Also, the comment above this line is now out-of-date.

**Comment by yhuai** [comment]
@JoshRosen I think you are right. It is not always safe to do this optimization. The result will be wrong whenever the cast causes information loss.

@viirya Sorry, I did not think about all cases before I opened the jira. How about we close this PR for now?

**Comment by viirya** [comment]
Sure. Close it now.
